loader/windows/icd_windows_dxgk.c

```diff
@@ -150,8 +150,8 @@ bool khrIcdOsVendorsEnumerateDXGK(void)
                 memset(cszLibraryName, 0, sizeof(cszLibraryName));
                 {
                     size_t len;
-                    wcstombs_s(&len, cszLibraryName, sizeof(cszLibraryName), pWchar, sizeof(cszLibraryName));
-                    KHR_ICD_ASSERT(len == (sizeof(cszLibraryName) - 1));
+                    wcstombs_s(&len, cszLibraryName, sizeof(cszLibraryName), pWchar, sizeof(cszLibraryName) - 1);
+                    KHR_ICD_ASSERT(len == wcslen(pWchar) + 1);
```

Comment:
@bashbaug please bear with me here, I am not a Windows expert, but I have read the documentation and have a nagging doubt. I am unsure this can work since (emphasis mine):

> **`wcstombs_s(size_t *pReturnValue, char *mbstr, size_t sizeInBytes, const wchar_t *wcstr, size_t count);`**
> `pReturnValue`
> The size in **bytes** of the converted string, including the null terminator.

> **`wcslen`**
> 1) Returns the length of a wide string, that is the **number** of non-null wide characters that precede the terminating null wide character.

So we're comparing bytes (as string which may include multi-bytes characters) to string length on wide characters. So if we have a wide character (something that results in a multibytes character) in the mix shouldn't that make `len` bigger and the assert fail, even if conversion was correct and successful?

Am I missing something here?

Reply:
Yeah, full disclosure: I'm not 100% sure if this is the right assert either, but here's how the thinking goes (and the assert does pass, at least in my limited testing):

If each character in the wide-character string pWchar gets converted to a single-byte character in the multi-byte string cszLibraryName then the assert is fine and it will pass.  This is probably the common case... but I suppose it isn't guaranteed either, hrm.

If we're not comfortable with this assumption I suppose we could call [_mbslen](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/strlen-wcslen-mbslen-mbslen-l-mbstrlen-mbstrlen-l?view=msvc-170) instead , which will return the number of characters in the multi-byte string, so (`KHR_ICD_ASSERT(_mbslen(cszLibraryName) == wcslen(pWchar))`).  Although, I am a little nervous about the "important note" in the docs:

> _mbslen, _mbslen_l, _mbstrlen, and _mbstrlen_l cannot be used in applications that execute in the Windows Runtime.

I'm a bit at a loss because I'm trying to infer what this assert is trying to test for.  Should we be testing for a successful return value from wctombs_s instead (non-zero)?  Is any assert required at all?

Reply:
I agree that testing success is the only thing we should do.

Reply:
Please take a look a the latest changes:

* I changed from a call to wctombs_s to a call to WideCharToMultiByte.  This matches behavior elsewhere in the OpenCL ICD loader and is consistent with the call to MultiByteToWideChar earlier in this function.
* I changed from a KHR_ICD_ASSERT to a check for success and a KHR_ICD_TRACE if the conversion failed, which is also consistent with the rest of the OpenCL ICD loader.

Reply:
Changes look great!

I am not a security expert, so I am not sure about this paragraph from the documentation (emphasis mine):

`[in] dwFlags`
`WC_NO_BEST_FIT_CHARS`
> Translate any Unicode characters that do not translate directly to multibyte equivalents to the default character specified by lpDefaultChar.  In other words, if translating from Unicode to multibyte and back to  Unicode again does not yield the same Unicode character, the function  uses the default character. This flag can be used by itself or in  combination with the other defined flags. For **strings that require validation, such as file**, resource, and user  names, the application should always use the WC_NO_BEST_FIT_CHARS flag.  This flag prevents the function from mapping characters to characters  that appear similar but have very different semantics. In some cases,  the semantic change can be extreme. For example, the symbol for "∞"  (infinity) maps to 8 (eight) in some code pages.

No idea if this is relevant in our context.

Reply:
> No idea if this is relevant in our context.

Yeah, I'm not sure either.  I did a quick grep through the OpenCL ICD loader and the Vulkan loader and they both are currently passing zero for `dwFlags`, so they do not include `WC_NO_BEST_FIT_CHARS`.

@jenatali do you have any experience in this area?

One delta I did notice between the OpenCL ICD loader and the Vulkan loader is that we are currently passing `CP_ACP` as our `CodePage` whereas Vulkan is consistently passing `CP_UTF8`, so this would be good to confirm also.

Reply:
I've got pretty minimal unicode experience, personally, and I've never used `WC_NO_BEST_FIT_CHARS` before.

As for `CP_ACP` vs `CP_UTF8`, you probably want `UTF8` unless there's a reason you want ASCII only.

Reply:
Thanks, I've switched from `CP_ACP` to `CP_UTF8` and my minimal testing still seems to be passing.

Reply:
@jenatali can we consider the issue resolved? I think with your approval we can merge.